lib/WTSI/NPG/HTS/Illumina/RunPublisher.pm

```diff
@@ -351,6 +364,47 @@ sub publish_alignment_files {
        with_spiked_control => $with_spiked_control);
   };
 
+  my $mlwh_json_cb = sub {
+    my ($obj, $file) = @_;
+    if (self->mlwh_json) {
+      my $mlwh_hash = {
+        id_product               => $obj->composition->digest,
+        seq_platform_name        => $ILLUMINA,
+        pipeline_name            => defined($self->alt_process) ?
+          $ALT_PROCESS : $NPG_PROD,
+        irods_root_collection    => $self->dest_collection,
+        irods_data_relative_path => $file,
+      };
+
+      open my $json_fh, '+<:encoding(UTF-8)', $self->mlwh_json or
+        self->logcroak(qq[could not open ml warehouse json file $self->mlwh_json]);
+      my $json_hash;
+      if (-e $self->mlwh_json) {
+        $json_hash = decode_json <$json_fh>;
+      }
+      else {
+        $json_hash = {
+          version  => $JSON_FILE_VERSION,
+          products => [],
+        };
+      }
+      push @{$json_hash->{products}}, $mlwh_hash;
+
+      seek $json_fh, 0, 0;
```

Comment:
This will seek to the start of the file and the next write will overwrite what is there. Is that what you mean?

```
$ touch test.txt
$ perl -e 'open my $f, "+<:encoding(UTF-8)", "test.txt" or die; print $f "hello\n"; seek $f, 0, 0; print $f "foo"; close $f or die'
$ cat test.txt
foolo
```

Reply:
Yes, I'm reading the file into memory, adding this product to the json object, then overwriting the original object, otherwise there would be multiple objects in the file, which I would prefer to avoid.

Reply:
Will the new one always be at least as long as the previous one? If not, you'll have trailing remnants of the previous.

Reply:
I had planned to move the file before using it to load the db, so that a new one could be created and not cause any confusion about what had already been loaded

Reply:
I'm going to merge this to progress things. I still think this operation is better done at a wider scope than this (we have precedent in the way re-start JSON files are created). The real test is how it performs, so we can use that criterion to evaluate the approach.